Add Integration tests to Auditbeat CI pipeline

mjwolf opened the pull request:
<!-- Type of change
Please label this PR with one of the following labels, depending on the scope of your change:
- Bug
- Enhancement
- Breaking change
- Deprecation
- Cleanup
- Docs
-->

## Proposed commit message

Refactor the CI integration test steps for the Auditbeat pipeline to use build matrices for Linux tests. By doing this, it will be easier to add additional VM types to the integration tests in the future.

With this, the integration tests will run on  Ubuntu 20.04, 22.04, 24.04 for both x86_64 and arm64, and RHEL 9 x86_64.

## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [x] My code follows the style guidelines of this project
- [ ] ~~I have commented my code, particularly in hard-to-understand areas~~
- [ ] ~~I have made corresponding changes to the documentation~~
- [ ] ~~I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] ~~I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.~~

elasticmachine commented:
## :broken_heart: Build Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-37892/detail/PR-37892/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-37892/detail/PR-37892/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-37892/detail/PR-37892/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-37892/detail/PR-37892/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://beats_37892.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/beats-ci/transactions/view?rangeFrom=2024-02-06T18:53:09.781Z&rangeTo=2024-02-06T19:13:09.781Z&transactionName=BUILD+Beats%2Fbeats%2FPR-%7Bnumber%7D&transactionType=job&latencyAggregationType=avg&traceId=0653fc014c419b7586f70653aab72681&transactionId=13686934c5faec52)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats





* Duration: 7 min 36 sec





</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->














### Pipeline error [![1](https://img.shields.io/badge/1%20-red)](https://beats-ci.elastic.co/job/Beats/job/beats/job/PR-37892/1//console)

> This error is likely related to the pipeline itself. Click <a href="https://beats-ci.elastic.co/job/Beats/job/beats/job/PR-37892/1/console">here</a>
> and then you will see the error (either incorrect syntax or an invalid configuration).



## :grey_exclamation: Flaky test report
No test was executed to be analysed.



## :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.


- `/package` : Generate the packages and run the E2E tests.


- `/beats-tester` : Run the installation tests with beats-tester.


- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)




</p>
</details>

<!--COMMENT_GENERATED_WITH_ID_comment.id-->

elasticmachine commented:
Pinging @elastic/sec-deployment-and-devices (Team:Security-Deployment and Devices)

mmat11 commented on .buildkite/auditbeat/auditbeat-pipeline.yml:
```diff
@@ -3,10 +3,14 @@
 env:
   IMAGE_UBUNTU_X86_64: "family/core-ubuntu-2204"
   IMAGE_UBUNTU_ARM_64: "core-ubuntu-2004-aarch64"
```
which kernel version does this ship with?

mjwolf replied:
It's using `5.15.0-1052-aws`

oakrizan replied:
IMAGE_UBUNTU_X86_64: Linux bk-agent-prod-gcp-1707303793816829620 6.2.0-1019-gcp #21~22.04.1-Ubuntu SMP Thu Nov 16 18:18:34 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
IMAGE_UBUNTU_ARM_64: Linux i-095e3c2e7d81fa82d 5.15.0-1052-aws #57~20.04.1-Ubuntu SMP Mon Jan 15 17:04:12 UTC 2024 aarch64 aarch64 aarch64 GNU/Linux

mmat11 commented on .buildkite/auditbeat/scripts/integ-tests.sh:
```diff
@@ -0,0 +1,12 @@
+sudo chmod -R go-w auditbeat/
+
+cd auditbeat
+umask 0022
```
why is this needed?

mjwolf replied:
It's not needed, I've removed it. I took most of this script from the unit test script, but this isn't needed here

oakrizan replied:
Buildkite-agent user has default umask=0002, meanwhile Jenkins has 0022. Without changing umask to 0022 tests are failing.

mmat11 commented:
I guess these can be removed then https://github.com/elastic/beats/blob/main/auditbeat/Jenkinsfile.yml#L33-L50

mmat11 commented on .buildkite/auditbeat/auditbeat-pipeline.yml:
```diff
@@ -105,6 +109,49 @@ steps:
+        command: ".buildkite/auditbeat/scripts/integ-tests.sh"
+        notify:
+          - github_commit_status:
+              context: "auditbeat/Extended: Linux Integration Tests"
```
nit: use `auditbeat: Integration Tests`, for consistency?

mjwolf replied:
I've changed to this name

mmat11 commented on .buildkite/auditbeat/auditbeat-pipeline.yml:
```diff
@@ -105,6 +109,49 @@ steps:
+        command: ".buildkite/auditbeat/scripts/integ-tests.sh"
+        notify:
+          - github_commit_status:
+              context: "auditbeat/Extended: Linux Integration Tests"
```
nit: use `auditbeat: Integration Tests`, for consistency?

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/filebeat/builds/2636)
* Commit: 6e0544b558defe4977d57fd541d503921a404213

### Failed CI Steps
* [:ubuntu: Python Integration Tests](https://buildkite.com/elastic/filebeat/builds/2636#018d80bf-dbcb-445d-a2bf-347b6fb073b5)

### History
* :broken_heart: [Build #2620](https://buildkite.com/elastic/filebeat/builds/2620) failed d98e7d09bc12c9db7d5002ace8fc042af7d83ba2

cc @mjwolf

<!--buildkite-pr-comment-filebeat
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/filebeat/builds/2636","number":2636,"commit":"6e0544b558defe4977d57fd541d503921a404213"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/filebeat/builds/2620","number":2620,"commit":"d98e7d09bc12c9db7d5002ace8fc042af7d83ba2"}],"number":2636}
buildkite-pr-comment-->

pkoutsovasilis commented on .buildkite/auditbeat/auditbeat-pipeline.yml:
```diff
@@ -1,12 +1,18 @@
+  IMAGE_DEBIAN_10: "family/core-debian-10"
+  IMAGE_DEBIAN_12: "family/core-debian-12"
+  IMAGE_CENTOS7: "family/core-centos-7"
+  IMAGE_SUSE15: "family/core-sles-15"
```
maybe also add `family/core-sles-12` it is in the support matrix of `auditbeat`

oakrizan replied:
[Same](https://github.com/elastic/beats/pull/37892#discussion_r1481285263)

pkoutsovasilis commented on .buildkite/auditbeat/auditbeat-pipeline.yml:
```diff
@@ -1,12 +1,18 @@
 env:
-  IMAGE_UBUNTU_X86_64: "family/core-ubuntu-2204"
-  IMAGE_UBUNTU_ARM_64: "core-ubuntu-2004-aarch64"
+  IMAGE_UBUNTU_2004_X86_64: "family/core-ubuntu-2004"
```
maybe also add `family/core-ubuntu-1804` it is in the support matrix of auditbeat

oakrizan commented on .buildkite/auditbeat/auditbeat-pipeline.yml:
```diff
@@ -1,12 +1,18 @@
+  IMAGE_DEBIAN_12: "family/core-debian-12"
+  IMAGE_CENTOS7: "family/core-centos-7"
+  IMAGE_SUSE15: "family/core-sles-15"
+  IMAGE_RHEL9: "family/core-rhel-9"
```
We are in process of migration beats from Jenkins to Buildkite, at the moment we are focusing on those agents, that were defined in Jenkinsfile and is available on Buildkite. Agent will be refined and added at the later stage, once all beats will be migrated to BK.

oakrizan commented:
> I guess these can be removed then https://github.com/elastic/beats/blob/main/auditbeat/Jenkinsfile.yml#L33-L50

Please keep the Jenkinsfiles in its' original state, since we still will need them

oakrizan reviewed (commented):
All the improvements/new features have to be postponed until the migration is completed. Migration has a hard deadline and is ongoing.

mjwolf commented:
> All the improvements/new features have to be postponed until the migration is completed. Migration has a hard deadline and is ongoing. cc: @dliappis @jlind23

I'll wait until the migration is completed before I merge this

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/filebeat/builds/2714)
* Commit: a7493ed507fdf645022f286d2aa4a78079dcb48d

### Failed CI Steps
* [Package pipeline](https://buildkite.com/elastic/filebeat/builds/2714#018d85aa-2c57-4873-958f-846c702226dc)

### History
* :broken_heart: [Build #2636](https://buildkite.com/elastic/filebeat/builds/2636) failed 6e0544b558defe4977d57fd541d503921a404213
* :broken_heart: [Build #2620](https://buildkite.com/elastic/filebeat/builds/2620) failed d98e7d09bc12c9db7d5002ace8fc042af7d83ba2

cc @mjwolf

<!--buildkite-pr-comment-filebeat
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/filebeat/builds/2714","number":2714,"commit":"a7493ed507fdf645022f286d2aa4a78079dcb48d"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/filebeat/builds/2636","number":2636,"commit":"6e0544b558defe4977d57fd541d503921a404213"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/filebeat/builds/2620","number":2620,"commit":"d98e7d09bc12c9db7d5002ace8fc042af7d83ba2"}],"number":2714}
buildkite-pr-comment-->

andrewkroh commented:
> I'll wait until the migration is completed before I merge this

I think buildkite the migration is mostly finished now. Is that correct, @dliappis?

So work on this could resume.

dliappis commented:
> I think buildkite the migration is mostly finished now. Is that correct, @dliappis?

Correct.

rowlandgeoff commented:
> either the VM image doesn't exist (Debian 10) or the image needs to be updated with additional tools to be able to run the tests (RHEL 8, SUSE 12). The images will need to be created/updated before they can be added to the integration tests.

Creating/updating these images will require planning and work on the Ingest Eng Prod team, and isn't something we can address during this PR. If you'd like to move forward with increasing your test coverage, I'd recommend pulling out the RHEL 8 and SLES 12 references until we can get to those. Assessing all of the beats and their test coverage relative to the compat matrix is something we were recently talking about, but it's going to take some time to assess and scope the work.

mjwolf commented:
> > either the VM image doesn't exist (Debian 10) or the image needs to be updated with additional tools to be able to run the tests (RHEL 8, SUSE 12). The images will need to be created/updated before they can be added to the integration tests.
> 
> Creating/updating these images will require planning and work on the Ingest Eng Prod team, and isn't something we can address during this PR. If you'd like to move forward with increasing your test coverage, I'd recommend pulling out the RHEL 8 and SLES 12 references until we can get to those. Assessing all of the beats and their test coverage relative to the compat matrix is something we were recently talking about, but it's going to take some time to assess and scope the work.

Hi, @rowlandgeoff, thanks, I see now that a lot of the images in the `platform-ingest-beats` family don't exist.  I was planning to add them myself (at least the easy ones that can be based off the existing core images), but if your team is planning to do it, I can wait for it.

For this PR, I've removed the images that don't exist and updated the description with what I've done now

pkoutsovasilis reviewed (approved):
ty @mjwolf this LGTM! such a distros extension of running Auditbeat FIM integration tests is critical as we utilise kprobes and ebpf technologies which can be affected with kernel changes and updates. @rowlandgeoff eager to see this getting more extended to include more distros from the Elastic compatibility matrix 🙂

dliappis commented on .buildkite/auditbeat/auditbeat-pipeline.yml:
```diff
@@ -128,49 +131,60 @@ steps:
         notify:
           - github_commit_status:
-              context: "auditbeat: Linux Integration Tests"
+              context: "auditbeat: Linux x86_64 Integration Tests"
```
By using a matrix and having a static context we'll be facing the issue that all failures will be collapsed under the same GH status notification.

For example, right now, with passing tests, the notification looks like:

![image](https://github.com/elastic/beats/assets/1754575/d40f4ea9-a1b5-4387-92f6-9cb7b29a496f)

If we click the link it takes us to:

![image](https://github.com/elastic/beats/assets/1754575/5be97e48-a102-4e77-890b-58e805a3dafd)

which represents both `Auditbeat Linux x86_64 Integration Tests -- family/platform-ingest-beats-ubuntu-2004` and `Auditbeat Linux x86_64 Integration Tests -- family/platform-ingest-beats-ubuntu-2204`

So if there is a failure, clicking on the link will take the user to the top of the matrix section, not the specific failed job. Looking at the [official BK docs](https://buildkite.com/docs/pipelines/build-matrix) it may not be possible to add `{{matrix.image}}` -- or some additional property lime `{{matrix.name}}` -- under `notify: ["github_commit_stats": {"context": "the message ... {{matrix.property}}"}]` (maybe worth testing to be 100% sure).

In any case, my main comment here is to raise awareness about the behavior and that with failures it won't be able to quickly see in the status section which specific image failed.

dliappis commented on .buildkite/auditbeat/auditbeat-pipeline.yml:
```diff
@@ -2,23 +2,26 @@
+  AWS_IMAGE_UBUNTU_2004_ARM64: "platform-ingest-beats-ubuntu-2004-aarch64"
+  AWS_IMAGE_UBUNTU_2204_ARM64: "platform-ingest-beats-ubuntu-2204-aarch64"
+
+  IMAGE_MACOS_ARM64: "generic-13-ventura-arm"
```
nit: re: renaming `IMAGE_MACOS_ARM` to `IMAGE_MACOS_ARM64`, ideally we do this in a future PR where we change this as a convention for all the pipelines. For now, it might be ok to leave the name as is, given that there is no 32bit ARM architecture for macOS?

dliappis reviewed (commented):
Looks good, thank you!

Left a comment to raise awareness on GH status notifications and that by using a matrix, failures will be collapsed under one entry. If you are ok with this I won't object, just wanted to make sure the behavior is clearly understood.

elasticmachine commented:
Pinging @elastic/elastic-agent-data-plane (Team:Elastic-Agent-Data-Plane)

mjwolf commented:
I'd like to revive this PR.  I've added all the Linux 'platform-ingest-beats' VM images that currently exist and removed the variable name changes that weren't necessary. 

@leehinman, @dliappis, @pkoutsovasilis could you review this again, since it's changed since you last looked at it?

rowlandgeoff commented:
LGTM. @oakrizan can you please review? Any overlap with your current changes?

oakrizan commented:
We deliberately were avoiding matrix usage. In addition to what Dimitrios said about matrix - if one of the matrix-based steps fails, then particular GH check will hang in "Pending" state.

oakrizan reviewed (approved):
lgtm
